File: backend/src/services/memory.ts
import { promises as fs } from 'fs';
import path from 'path';
import type { MemoryData } from '../types';
import { getDataDir } from '../utils/helpers';

const MEMORY_DIR = path.join(getDataDir(), 'memory');

const DEFAULT_MEMORY: MemoryData = {
  enabled: false,
  categories: {},
  updatedAt: 0,
};

// ─── In-Memory Cache (#4) ──────────────────────────────────
// Cache memory data per-user for 5 seconds to avoid disk reads on every message
const MEMORY_CACHE_TTL = 5_000; // 5 seconds
const memoryCache = new Map<string, { data: MemoryData; timestamp: number }>();

function getCachedMemory(userId: string): MemoryData | null {
  const entry = memoryCache.get(userId);
  if (entry && Date.now() - entry.timestamp < MEMORY_CACHE_TTL) {
    return entry.data;
  }
  return null;
}

function setCachedMemory(userId: string, data: MemoryData): void {
  memoryCache.set(userId, { data, timestamp: Date.now() });
}

function invalidateCache(userId: string): void {
  memoryCache.delete(userId);
}

// Per-user write queue to prevent race conditions when multiple extractions run concurrently
const writeQueues = new Map<string, Promise<unknown>>();

async function serializedWrite<T>(userId: string, fn: () => Promise<T>): Promise<T> {
  const prev = (writeQueues.get(userId) || Promise.resolve()) as Promise<unknown>;
  const next = prev.then(fn, fn);
  writeQueues.set(userId, next);
  next.finally(() => {
    if (writeQueues.get(userId) === next) {
      writeQueues.delete(userId);
    }
  });
  return next as Promise<T>;
}

async function ensureDir(): Promise<void> {
  await fs.mkdir(MEMORY_DIR, { recursive: true });
}

function memoryFilePath(userId: string): string {
  const safe = userId.replace(/[^a-zA-Z0-9_\-]/g, '_');
  return path.join(MEMORY_DIR, `${safe}.json`);
}

export async function getMemory(userId: string): Promise<MemoryData> {
  // Check cache first
  const cached = getCachedMemory(userId);
  if (cached) return cached;

  try {
    await ensureDir();
    const filePath = memoryFilePath(userId);
    const data = await fs.readFile(filePath, 'utf-8');
    const result = { ...DEFAULT_MEMORY, ...JSON.parse(data) };
    setCachedMemory(userId, result);
    return result;
  } catch {
    return { ...DEFAULT_MEMORY };
  }
}

export async function saveMemory(userId: string, memory: MemoryData): Promise<MemoryData> {
  await ensureDir();
  const filePath = memoryFilePath(userId);
  const next: MemoryData = {
    ...memory,
    updatedAt: Date.now(),
  };
  await fs.writeFile(filePath, JSON.stringify(next, null, 2));
  invalidateCache(userId); // Invalidate cache on write
  setCachedMemory(userId, next); // Pre-populate cache
  return next;
}

export async function updateMemory(
  userId: string,
  updates: Partial<Pick<MemoryData, 'enabled'>> & {
    categories?: Record<string, Record<string, string>>;
  }
): Promise<MemoryData> {
  return serializedWrite(userId, async () => {
    const current = await getMemory(userId);
    const next: MemoryData = {
      ...current,
      ...updates,
      categories: updates.categories ?? current.categories,
      updatedAt: Date.now(),
    };
    return saveMemory(userId, next);
  });
}

export async function mergeMemoryEntries(
  userId: string,
  extracted: Record<string, Record<string, string>>
): Promise<MemoryData | null> {
  return serializedWrite(userId, async () => {
    const current = await getMemory(userId);
    if (!current.enabled) return null;

    const merged: Record<string, Record<string, string>> = { ...current.categories };

    for (const [category, entries] of Object.entries(extracted)) {
      if (!merged[category]) {
        const nonEmpty = Object.fromEntries(
          Object.entries(entries).filter(([, v]) => v !== '')
        );
        if (Object.keys(nonEmpty).length > 0) {
          merged[category] = nonEmpty;
        }
        continue;
      }

      const categoryEntries = { ...merged[category] };
      for (const [key, value] of Object.entries(entries)) {
        if (value === '') {
          delete categoryEntries[key];
        } else {
          categoryEntries[key] = value;
        }
      }

      if (Object.keys(categoryEntries).length === 0) {
        delete merged[category];
      } else {
        merged[category] = categoryEntries;
      }
    }

    const next: MemoryData = {
      ...current,
      enabled: Object.keys(merged).length > 0 ? true : current.enabled,
      categories: merged,
      updatedAt: Date.now(),
    };

    return saveMemory(userId, next);
  });
}

File: backend/app/memory.py
"""Memory service — mirrors backend/src/services/memory.ts.

Per-user memory stored in data/memory/<user>.json, 5s cache, serialized
writes per user to prevent race conditions between concurrent extractions.
"""

from __future__ import annotations

import asyncio
import json
import re
import time
from pathlib import Path
from typing import Any

from .config import get_data_dir
from .logger import error as log_error

MEMORY_CACHE_TTL = 5.0
_memory_cache: dict[str, tuple[dict[str, Any], float]] = {}
_write_locks: dict[str, asyncio.Lock] = {}

DEFAULT_MEMORY: dict[str, Any] = {"enabled": False, "categories": {}, "updatedAt": 0}


def _memory_dir() -> Path:
    return get_data_dir() / "memory"


def _memory_file_path(user_id: str) -> Path:
    safe = re.sub(r"[^a-zA-Z0-9_\-]", "_", user_id)
    return _memory_dir() / f"{safe}.json"


def _get_cached(user_id: str) -> dict[str, Any] | None:
    entry = _memory_cache.get(user_id)
    if entry and time.time() * 1000 - entry[1] < MEMORY_CACHE_TTL * 1000:
        return entry[0]
    return None


def _lock_for(user_id: str) -> asyncio.Lock:
    if user_id not in _write_locks:
        _write_locks[user_id] = asyncio.Lock()
    return _write_locks[user_id]


async def get_memory(user_id: str) -> dict[str, Any]:
    cached = _get_cached(user_id)
    if cached is not None:
        return cached
    try:
        with open(_memory_file_path(user_id), encoding="utf-8") as f:
            data = json.load(f)
        result = {**DEFAULT_MEMORY, **data}
    except FileNotFoundError:
        return dict(DEFAULT_MEMORY)
    except Exception as e:  # noqa: BLE001
        log_error("[memory] Load failed:", e)
        return dict(DEFAULT_MEMORY)
    _memory_cache[user_id] = (result, time.time() * 1000)
    return result


async def save_memory(user_id: str, memory: dict[str, Any]) -> dict[str, Any]:
    _memory_dir().mkdir(parents=True, exist_ok=True)
    nxt = {**memory, "updatedAt": int(time.time() * 1000)}
    try:
        with open(_memory_file_path(user_id), "w", encoding="utf-8") as f:
            json.dump(nxt, f, indent=2)
    except Exception as e:  # noqa: BLE001
        log_error("[memory] Save failed:", e)
        raise
    _memory_cache[user_id] = (nxt, time.time() * 1000)
    return nxt


async def update_memory(user_id: str, updates: dict[str, Any]) -> dict[str, Any]:
    async with _lock_for(user_id):
        current = await get_memory(user_id)
        nxt = {**current, **updates}
        nxt["categories"] = updates.get("categories", current.get("categories", {}))
        return await save_memory(user_id, nxt)


async def merge_memory_entries(
    user_id: str, extracted: dict[str, dict[str, str]]
) -> dict[str, Any] | None:
    async with _lock_for(user_id):
        current = await get_memory(user_id)
        if not current.get("enabled"):
            return None

        merged: dict[str, dict[str, str]] = dict(current.get("categories", {}))

        for category, entries in extracted.items():
            if category not in merged:
                non_empty = {k: v for k, v in entries.items() if v != ""}
                if non_empty:
                    merged[category] = non_empty
                continue

            category_entries = dict(merged[category])
            for key, value in entries.items():
                if value == "":
                    category_entries.pop(key, None)
                else:
                    category_entries[key] = value
            if not category_entries:
                del merged[category]
            else:
                merged[category] = category_entries

        nxt = {
            **current,
            "enabled": True if merged else current.get("enabled", False),
            "categories": merged,
            "updatedAt": int(time.time() * 1000),
        }
        return await save_memory(user_id, nxt)
